iothub/device/src/ClientOptions.cs

```diff
@@ -29,5 +30,27 @@ public class ClientOptions
         /// The default behavior is that <see cref="Message.MessageId"/> is set only by the user.
         /// </summary>
         public SdkAssignsMessageId SdkAssignsMessageId { get; set; } = SdkAssignsMessageId.Never;
+
+        /// <summary>
+        /// The suggested time to live value for tokens generated for SAS authenticated clients.
+        /// If unset the generated SAS tokens will be valid for 1 hour.
+        /// </summary>
+        /// <remarks>
+        /// This is used only for SAS token authenticated clients through either the
+        /// <see cref="DeviceClient.CreateFromConnectionString(string, ClientOptions)"/>/ <see cref="ModuleClient.CreateFromConnectionString(string, ClientOptions)"/> flow
+        /// or the <see cref="ModuleClient.CreateFromEnvironmentAsync(ClientOptions)"/> flow.
+        /// </remarks>
+        public TimeSpan SasTokenTimeToLive { get; set; }
+
+        /// <summary>
+        /// The time buffer before expiry when the token should be renewed, expressed as a percentage of the time to live.
+        /// If unset the token will be renewed when it has 15% or less of its lifespan left.
+        /// </summary>
+        /// <remarks>
+        /// This is used only for SAS token authenticated clients through either the
+        /// <see cref="DeviceClient.CreateFromConnectionString(string, ClientOptions)"/>/ <see cref="ModuleClient.CreateFromConnectionString(string, ClientOptions)"/> flow
+        /// or the <see cref="ModuleClient.CreateFromEnvironmentAsync(ClientOptions)"/> flow.
+        /// </remarks>
+        public int SasTokenRenewalBuffer { get; set; }
```

Comment:
An int can be way bigger than 100 and way smaller than 1. Are we going to do any value checking?

Same question for TTL.

Reply:
We do have these validations in place in the base class where these values are used: https://github.com/Azure/azure-iot-sdk-csharp/blob/master/iothub/device/src/AuthenticationWithTokenRefresh.cs#L36-L44

Reply:
Yeah, but well after it is set. Shouldn't we give the user feedback immediately that the value is invalid?

Reply:
My idea was to keep the validation in a single location; i.e. you could set the token ttl through multiple locations, but they'd all get set to AuthenticationWithTokenRefresh finally, and that is where the validation would happen. This would help minimize chances of us implementing different validation rules in different places.
Since these values are set as a part of the client initialization process, I thought it was fine to delay it until it reaches AuthenticationWithTokenRefresh

Reply:
I can appreciate that, but how does a user get feedback about the validity of their values?

Reply:
They'll get an ArgumentOutOfRangeException when they create the client instance. I've added some unit tests around this.

Reply:
That's not terrible. Would you at least put acceptable ranges in the doc comments for those properties?

Reply:
Sure thing.